src/Generation/GapicMetadataGenerator.php

```diff
@@ -0,0 +1,89 @@
+ * limitations under the License.
+ */
+declare(strict_types=1);
+
+namespace Google\Generator\Generation;
+
+use Google\Generator\Collections\Vector;
+use Google\Generator\Utils\ProtoCatalog;
+use Google\Gapic\Metadata\GapicMetadata;
+
+class GapicMetadataGenerator
+{
+    /**
+     *  Generates the contents of a gapic_metadata.json file.
+     */
+    public static function generate(ProtoCatalog $catalog, Vector $fileDescriptors, string $namespace): string
+    {
+        return (new GapicMetadataGenerator($catalog, $fileDescriptors, $namespace))->generateImpl();
+    }
+
+    private ProtoCatalog $catalog;
+    private Vector $fileDescriptors;
+    private string $namespace;
+    private GapicMetadata $gapicMetadata;
+
+    private function __construct(ProtoCatalog $catalog, Vector $fileDescriptors, string $namespace)
+    {
+        $this->catalog = $catalog;
+        $this->fileDescriptors = $fileDescriptors;
+        $this->namespace = $namespace;
+        $this->gapicMetadata = new GapicMetadata();
+    }
+
+    private function generateImpl(): string
+    {
+        // Initialization.
+        $this->gapicMetadata->setSchema("1.0");
+        $this->gapicMetadata->setComment("This file maps proto services/RPCs to the corresponding library clients/methods");
+        $this->gapicMetadata->setLanguage("php");
+        $this->gapicMetadata->setLibraryPackage($this->namespace);
+
+        $gapicMetadataServices = [];
+        // TODO: Filter out mixed-in services.
+        foreach ($this->fileDescriptors as $fileDesc) {
+            // Only need to set this once.
+            if (empty($this->gapicMetadata->getProtoPackage)) {
+                $this->gapicMetadata->setProtoPackage($fileDesc->getPackage());
+            }
+            foreach ($fileDesc->getService() as $index => $service) {
+                $serviceName = "{$fileDesc->getPackage()}.{$service->getName()}";
+                $serviceDetails =
+          new ServiceDetails($this->catalog, $this->namespace, $fileDesc->getPackage(), $service, $fileDesc);
+                $gapicMetadataClients = [];
+                $rpcs = $serviceDetails->methods->toArray(
+                    fn ($method) => $method->name,
+                    fn ($method) => [$method->methodName]
+                );
+
+                $libraryClient = new GapicMetadata\ServiceAsClient();
+                $libraryClient->setLibraryClient($serviceDetails->gapicClientType->name);
+                $libraryClient->setRpcs(
+                    $serviceDetails->methods->toArray(
+                        fn ($method) => $method->name,
+                        fn ($method) => (new GapicMetadata\MethodList())->setMethods([$method->methodName])
+                    )
+                );
+                $transport = new GapicMetadata\ServiceForTransport();
+                $transport->setClients(['grpc' => $libraryClient]);
+                $gapicMetadataServices[$serviceDetails->serviceName] = $transport;
```

Comment:
This should NOT be the fully qualified service name, but rather the part after the value in `protoPackage` (see my other comment)

Reply:
Done.